chain-extensions/pallet-assets/src/lib.rs

```diff
@@ -0,0 +1,411 @@
+            10 => Ok(AssetsFunc::TransferApproved),
+            11 => Ok(AssetsFunc::SetMetadata),
+            12 => Ok(AssetsFunc::MetadataName),
+            13 => Ok(AssetsFunc::MetadataSymbol),
+            14 => Ok(AssetsFunc::MetadataDecimals),
+            _ => Err(DispatchError::Other(
+                "PalletAssetsExtension: Unimplemented func_id",
+            )),
+        }
+    }
+}
+
+/// Pallet Assets chain extension.
+pub struct AssetsExtension<T>(PhantomData<T>);
+
+impl<T> Default for AssetsExtension<T> {
+    fn default() -> Self {
+        AssetsExtension(PhantomData)
+    }
+}
+
+impl<T> ChainExtension<T> for AssetsExtension<T>
+where
+    T: pallet_assets::Config + pallet_contracts::Config,
+    <<T as SysConfig>::Lookup as StaticLookup>::Source: From<<T as SysConfig>::AccountId>,
+    <T as SysConfig>::AccountId: From<[u8; 32]>,
+{
+    fn call<E: Ext>(&mut self, env: Environment<E, InitState>) -> Result<RetVal, DispatchError>
+    where
+        E: Ext<T = T>,
+        <E::T as SysConfig>::AccountId: UncheckedFrom<<E::T as SysConfig>::Hash> + AsRef<[u8]>,
+    {
+        let func_id = env.func_id().try_into()?;
+        let mut env = env.buf_in_buf_out();
+
+        match func_id {
+            AssetsFunc::Create => {
+                let (origin, id, admin, min_balance): (
+                    Origin,
+                    <T as pallet_assets::Config>::AssetId,
+                    T::AccountId,
+                    T::Balance,
+                ) = env.read_as()?;
+
+                let base_weight = <T as pallet_assets::Config>::WeightInfo::create();
+                env.charge_weight(base_weight)?;
+
+                let runtime_origin = RawOrigin::Signed(match origin {
+                    Origin::Caller => env.ext().caller().clone(),
+                    Origin::Address => env.ext().address().clone(),
+                });
+
+                let call_result = pallet_assets::Pallet::<T>::create(
+                    runtime_origin.into(),
+                    id.into(),
+                    admin.into(),
+                    min_balance,
+                );
+                return match call_result {
+                    Err(e) => {
+                        let mapped_error = Outcome::from(e);
+                        Ok(RetVal::Converging(mapped_error as u32))
+                    }
+                    Ok(_) => Ok(RetVal::Converging(Outcome::Success as u32)),
+                };
+            }
+            AssetsFunc::Transfer => {
+                let (origin, id, target, amount): (
+                    Origin,
+                    <T as pallet_assets::Config>::AssetId,
+                    T::AccountId,
+                    T::Balance,
+                ) = env.read_as()?;
+
+                let base_weight = <T as pallet_assets::Config>::WeightInfo::transfer();
+                env.charge_weight(base_weight)?;
+
+                let runtime_origin = RawOrigin::Signed(match origin {
```

Comment:
May be having some kind of template or function for this match could be useful for improving readability.